# Linear Regression

In [1]:
import numpy as np
from sklearn.datasets import load_boston
X = load_boston()

In [2]:
# split the dataset into train and test
from sklearn.cross_validation import train_test_split
X_tr, X_te, y_tr, y_te = train_test_split(X.data, X.target)

In [3]:
m = len(X_tr)
x0 = np.ones((m,1))
X = np.hstack((x0,X_tr))
Y = np.array(y_tr)
B = np.array(np.zeros(14))
alpha = 1e-6

In [4]:
def cost_function(X,Y,B):
    m = len(Y)
    J = np.sum((np.matmul(X,B) - Y)**2) / (2*m)
    return J

In [5]:
def Grad_descent(X,Y,B,alpha,iterations):
    cost = [0]*iterations
    m = len(Y)
    for i in range(iterations):
        hypo = np.matmul(X,B)
        loss = hypo - Y
        grad = np.matmul(loss,X)/m
        B = B - (alpha * grad)
        J = cost_function(X, Y, B)
        cost[i] = J
    return B, cost
        

In [6]:
Icost = cost_function(X,Y,B)
finalB, finalCost = Grad_descent(X,Y,B,alpha,100000)

In [7]:
def rmse(Y_actual, Y_predicted):
    error = np.sqrt(sum((Y_actual - Y_predicted) ** 2) / len(Y_actual))
    return error

n = len(X_te)
x0 = np.ones((n,1))
X_te = (np.hstack((x0,X_te)))

Y_predicted = np.matmul(X_te,finalB)

print (" Error := ", rmse(y_te,Y_predicted))

 Error :=  7.4644886898


# Observations -

|# Iterations | lr|RMSE|
|:-------:|:------:|:-----:|:----:|
| 100000| 1e-6 | 7.464|
| 100000| 1e-7 | 9.005 |
| 10000| 1e-6 | 8.284|
| 10000| 1e-7 | 8.344 |
|1000| 1e-6 | 8.727|
|1000| 1e-7 | 8.303 |

For a learning rate above 1e-6, the error is unbounded, hence it is assumed to be very low.

For a gradient descent function which is being used to reduce the loss incurred by the model, the error term oscillates.

The optimum(minimum) error is obtained when the model is run for 100000 iterations with a learning rate of 1e-6, with an error rate of 7.46 which is the next best to the actual error obtained when run using the sklearn.linear_model() function On the boston housing dataset.